### PR TITLE
[CodeGen] Convert depthwise mhlo.convolution to Linalg counterpart

### PR DESCRIPTION
**Pull request by antiagainst**
This commit adds a pattern to convert depthwise mhlo.convolution
to linalg.depthwise_conv_2d_nhwc when possible. The pattern is
given a high priority so it can take precedence.

### CODE REVIEW

**Comment by antiagainst** [comment]
This will fail for now because it need llvm/llvm-project@068bf9e to be pulled in.

File: iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnBuffers.cpp [hunk 1]
```diff
@@ -497,6 +526,89 @@ LogicalResult ConvOpConversion::apply(
...
+
+  if (failed(zeroFillBuffer(op.getLoc(), resultBuffers[0], rewriter))) {
+    return rewriter.notifyMatchFailure(op, "failed to zero fill result buffer");
+  }
+
+  // Create a Linalg reshape op that converts the filter from 4 dimensions
+  // into 3 dimensions (by droping the unit dimension). This is needed because
+  // linalg.depthwise_conv_2d_nhwc expects 3 dimensions for the filter.
+
+  auto outputDims =
```
Comment (asaadaldien) [inline iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnBuffers.cpp:582]: nit:FilterDims

File: iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnBuffers.cpp [hunk 2]
```diff
@@ -497,6 +526,89 @@ LogicalResult ConvOpConversion::apply(
...
+      llvm::to_vector<4>(op.rhs().getType().cast<ShapedType>().getShape());
+  if (outputDims[2] * outputDims[3] != op.feature_group_count()) {
+    return rewriter.notifyMatchFailure(
+        op, "non-one channel multiplier unsupported yet");
+  }
+  outputDims[2] = op.feature_group_count();
+  outputDims.pop_back();
+
+  MemRefType outputShape = MemRefType::get(
```
Comment (asaadaldien) [inline iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnBuffers.cpp:591]: nit: filterShape

File: iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnBuffers.cpp [hunk 3]
```diff
@@ -497,6 +526,89 @@ LogicalResult ConvOpConversion::apply(
...
+
+  if ((op.lhs_dilation() && !isSplatValue(*op.lhs_dilation(), 1)) ||
+      (op.rhs_dilation() && !isSplatValue(*op.rhs_dilation(), 1))) {
+    return rewriter.notifyMatchFailure(op, "non-one dialation unsupported yet");
+  }
+
+  if (const mhlo::ConvDimensionNumbers &dimension_numbers =
+          op.dimension_numbers()) {
+    // Make sure that this is 2-D convolution.
+    const int spatial_rank =
```
Comment (hanhanW) [inline iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnBuffers.cpp:546]: spatialRank

File: iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnBuffers.cpp [hunk 4]
```diff
@@ -497,6 +526,89 @@ LogicalResult ConvOpConversion::apply(
...
+        llvm::size(dimension_numbers.input_spatial_dimensions());
+    if (spatial_rank != 2) {
+      return rewriter.notifyMatchFailure(op, "only support 2-D cases for now");
+    }
+
+    // Make sure that this is depthwise convolution.
+    int64_t input_feature_dim =
```
Comment (hanhanW) [inline iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnBuffers.cpp:553]: inputFeatureDim

File: iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnBuffers.cpp [hunk 5]
```diff
@@ -497,6 +526,89 @@ LogicalResult ConvOpConversion::apply(
...
+        dimension_numbers.input_feature_dimension().getInt();
+    int64_t input_feature_count =
```
Comment (hanhanW) [inline iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnBuffers.cpp:555]: inputFeatureCount

**Review by asaadaldien — APPROVED** [review]

**Pull request merged**